examples/weakref.cpp

```diff
@@ -0,0 +1,301 @@
+                                        [[maybe_unused]],
+                                        void* user_data) {
+  // Queue a cleanup task to run after each microtask has been run.
+  // We only have one global so ignore the incumbent global parameter.
+  auto* jobQueue = static_cast<CustomJobQueue*>(user_data);
+  jobQueue->queueFinalizationRegistryCallback(callback);
+}
+
+static bool GC(JSContext* cx, unsigned argc, JS::Value* vp) {
+  JS::CallArgs args = JS::CallArgsFromVp(argc, vp);
+
+  JS_GC(cx, JS::GCReason::API);
+
+  args.rval().setUndefined();
+  return true;
+}
+
+static bool RunJobs(JSContext* cx, unsigned argc, JS::Value* vp) {
+  JS::CallArgs args = JS::CallArgsFromVp(argc, vp);
+
+  // This calls JS::ClearKeptObjects() after draining the microtask queue. If
+  // you're not using js::RunJobs(), you'll have to call it yourself --
+  // otherwise, the WeakRefs will never be emptied.
+  js::RunJobs(cx);
+
+  args.rval().setUndefined();
+  return true;
+}
+
+static bool ExecuteCode(JSContext* cx, const char* code) {
+  JS::CompileOptions options{cx};
+  options.setFileAndLine("noname", 1);
+
+  JS::SourceText<mozilla::Utf8Unit> source;
+  if (!source.init(cx, code, strlen(code), JS::SourceOwnership::Borrowed)) {
+    return false;
+  }
+
+  JS::Rooted<JS::Value> rval{cx};
+  return JS::Evaluate(cx, options, source, &rval);
+}
+
+static bool WeakRefExample(JSContext* cx) {
+  // Using WeakRefs and FinalizationRegistry requires a job queue. The built-in
+  // job queue used in repl.cpp is not sufficient, because it does not provide
+  // any way to queue FinalizationRegistry cleanup callbacks.
+  CustomJobQueue jobQueue{cx};
+  JS::SetJobQueue(cx, &jobQueue);
+
+  // Without this, FinalizationRegistry callbacks will never be called. The
+  // embedding has to decide when to schedule them.
+  JS::SetHostCleanupFinalizationRegistryCallback(
+      cx, CleanupFinalizationRegistry, &jobQueue);
+
+  JS::RealmOptions options;
+  options.creationOptions().setWeakRefsEnabled(
+      JS::WeakRefSpecifier::EnabledWithoutCleanupSome);
+
+  static JSClass GlobalClass = {"WeakRefsGlobal", JSCLASS_GLOBAL_FLAGS,
+                                &JS::DefaultGlobalClassOps};
+
+  JS::Rooted<JSObject*> global{
+      cx, JS_NewGlobalObject(cx, &GlobalClass, nullptr, JS::FireOnNewGlobalHook,
+                             options)};
+  if (!global) return false;
+
+  JSAutoRealm ar{cx, global};
+
+  if (!JS_DefineFunction(cx, global, "gc", &GC, 0, 0) ||
+      !JS_DefineFunction(cx, global, "runJobs", &RunJobs, 0, 0)) {
+    boilerplate::ReportAndClearException(cx);
+    return false;
+  }
+
+  if (!ExecuteCode(cx, R"js(
+    let valueFinalized;
+    const registry = new FinalizationRegistry(
+      heldValue => (valueFinalized = heldValue));
+    let obj = {};
+    const weakRef = new WeakRef(obj);
+    registry.register(obj, "marker");
+
+    obj = null;
+
+    runJobs();  // Clears weakRef. obj is now eligible for collection
```

Comment:
This doesn't actually clear `weakRef`.

This makes `obj` (actually, the object once pointed by `obj`) not "live", and `weakRef` is cleared as part of GC, given the pointed object is not "live" at that point.  if you put something about `weakRef`, it could be "weakRef is now eligible for clear" or something?

Details in the following:

After `obj = null`, the object is referred only from `agent.[[KeptAlive]]` and `weakRef`.
Given `weakRef` doesn't make it "live", only `agent.[[KeptAlive]]` makes it "live".

https://tc39.es/ecma262/#sec-liveness
> At any point during evaluation, a set of objects and/or symbols S is considered live if either of the following conditions is met:
>    * Any element in S is included in any [agent](https://tc39.es/ecma262/#agent)'s [[KeptAlive]] [List](https://tc39.es/ecma262/#sec-list-and-record-specification-type).

Then, `ClearKeptObjects` clears `agent.[[KeptAlive]]` list, which removes the last reference to the object.
so, "obj is now eligible for collection" part is true.

Then, the actual place where the `weakRef` is cleared is the following:

https://tc39.es/ecma262/#sec-weakref-execution
> At any time, if a set of objects and/or symbols S is not [live](https://tc39.es/ecma262/#sec-liveness), an ECMAScript implementation may perform the following steps atomically:
> 
> * 1. For each element value of S, do
>     * a. For each [WeakRef](https://tc39.es/ecma262/#sec-weak-ref-constructor) ref such that ref.[[WeakRefTarget]] is value, do
>         *  i. Set ref.[[WeakRefTarget]] to empty.

While the spec says "at any time", this is actually done by GC in the implementation.